File: Die.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace TableDieRangeCalculator
{
    public class Die
    {
        private static Dictionary<string, Die> _dice = new Dictionary<string, Die>();

        public static Die d2 = new Die(1, 2, nameof(d2));
        public static Die d3 = new Die(1, 3, nameof(d3));
        public static Die d4 = new Die(1, 4, nameof(d4));
        public static Die d6 = new Die(1, 6, nameof(d6));
        public static Die d8 = new Die(1, 8, nameof(d8));
        public static Die d10 = new Die(1, 10, nameof(d10));
        public static Die d12 = new Die(1, 12, nameof(d12));
        public static Die d16 = new Die(1, 16, nameof(d16));
        public static Die d20 = new Die(1, 20, nameof(d20));
        public static Die d24 = new Die(1, 24, nameof(d24));
        public static Die d30 = new Die(1, 30, nameof(d30));
        public static Die d100 = new Die(1, 100, nameof(d100));
        public static Die d1000 = new Die(1, 1000, nameof(d1000));
        
        public static Die GetDie(string name)
        {
            if (string.Equals(name, "d%", StringComparison.OrdinalIgnoreCase))
            {
                return d100;
            }
            else
            {
                return _dice[name];
            }
        }

        public Die(int sides)
        {
            MinRoll = 1;
            MaxRoll = sides;
            Name = $"d{sides}";
        }

        private Die(int minRoll, int maxRoll, string name)
        {
            MinRoll = minRoll; 
            MaxRoll = maxRoll;
            Name = name;
            _dice.Add(name, this);
        }

        public int MinRoll { get; }
        public int MaxRoll { get; }
        public string Name { get; }
        public int Sides => MaxRoll - MinRoll + 1;
        public IEnumerable<int> Faces => Enumerable.Range(MinRoll, MaxRoll);
        public decimal FaceProbability => 1m / Sides;
        
        public string GetHeader()
        {
            return GetHeader(MinRoll, MaxRoll);
        }

        public static string GetHeader(int minRoll, int maxRoll)
        {
            Die? found = _dice.Values.SingleOrDefault(d => d.MinRoll == minRoll && d.MaxRoll == maxRoll);
            if (found != null)
            {
                return found.Name + " Roll";
            }
            else if (minRoll == 1)
            {
                return "d" + maxRoll.ToString() + " Roll";
            }
            else
            {
                return "Roll";
            }
        }

        public static implicit operator Dice(Die die)
        {
            return new Dice(die);
        }
    }
}

File: node/src/die.js
'use strict';

const { Decimal } = require('./decimal');

const registry = new Map();

class Die {
  constructor(sides) {
    this.minRoll = 1;
    this.maxRoll = sides;
    this.name = `d${sides}`;
  }

  static register(minRoll, maxRoll, name) {
    const die = Object.create(Die.prototype);
    die.minRoll = minRoll;
    die.maxRoll = maxRoll;
    die.name = name;
    registry.set(name, die);
    return die;
  }

  static getDie(name) {
    if (name.toLowerCase() === 'd%') return Die.d100;
    const die = registry.get(name);
    if (die === undefined) {
      throw new Error(`The given key '${name}' was not present in the dictionary.`);
    }
    return die;
  }

  get sides() {
    return this.maxRoll - this.minRoll + 1;
  }

  get faceProbability() {
    return Decimal.ONE.divide(Decimal.fromInt(this.sides));
  }

  getHeader() {
    return Die.getHeaderFor(this.minRoll, this.maxRoll);
  }

  static getHeaderFor(minRoll, maxRoll) {
    const matches = [...registry.values()].filter((d) => d.minRoll === minRoll && d.maxRoll === maxRoll);
    if (matches.length === 1) return `${matches[0].name} Roll`;
    if (minRoll === 1) return `d${maxRoll} Roll`;
    return 'Roll';
  }
}

Die.d2 = Die.register(1, 2, 'd2');
Die.d3 = Die.register(1, 3, 'd3');
Die.d4 = Die.register(1, 4, 'd4');
Die.d6 = Die.register(1, 6, 'd6');
Die.d8 = Die.register(1, 8, 'd8');
Die.d10 = Die.register(1, 10, 'd10');
Die.d12 = Die.register(1, 12, 'd12');
Die.d16 = Die.register(1, 16, 'd16');
Die.d20 = Die.register(1, 20, 'd20');
Die.d24 = Die.register(1, 24, 'd24');
Die.d30 = Die.register(1, 30, 'd30');
Die.d100 = Die.register(1, 100, 'd100');
Die.d1000 = Die.register(1, 1000, 'd1000');

module.exports = { Die };
